homeassistant/components/rainmachine/config_flow.py

```diff
@@ -19,58 +21,129 @@
 )
 
 
+def get_client_controller(client):
+    """Enumerate controllers to find the first mac."""
+    for controller in client.controllers.values():
+        return controller
+
+
+async def async_get_controller(hass, ip_address, password, port, ssl):
+    """Auth and fetch the mac address from the controller."""
+    websession = aiohttp_client.async_get_clientsession(hass)
+    client = Client(session=websession)
+    try:
+        await client.load_local(ip_address, password, port=port, ssl=ssl)
+    except RainMachineError:
+        return None
+    else:
+        return get_client_controller(client)
+
+
 class RainMachineFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
     """Handle a RainMachine config flow."""
 
     VERSION = 1
 
+    def __init__(self):
+        """Initialize config flow."""
+        self.discovered_ip_address = None
+
     @staticmethod
     @callback
     def async_get_options_flow(config_entry):
         """Define the config flow to handle options."""
         return RainMachineOptionsFlowHandler(config_entry)
 
-    async def async_step_user(self, user_input=None):
-        """Handle the start of the config flow."""
-        if not user_input:
-            return self.async_show_form(
-                step_id="user", data_schema=DATA_SCHEMA, errors={}
-            )
-
-        await self.async_set_unique_id(user_input[CONF_IP_ADDRESS])
+    @callback
+    def _async_abort_ip_address_configured(self, ip_address):
+        """Abort if we already have an entry for the ip."""
+        # IP already configured
+        for entry in self._async_current_entries(include_ignore=False):
+            if ip_address in entry.data[CONF_IP_ADDRESS]:
```

Comment:
These are strings, use equality operator instead of searching for substring.

```suggestion
            if ip_address == entry.data[CONF_IP_ADDRESS]:
```

Reply:
Thanks.  I couldn't find where I copied that from but that's telling me we probably could use self._async_abort_entries_match_any({CONF_HOST: "1.2.3.4"}). Since nearly every integration that does local discovery has this pattern and it's written out 100x of times in the code base.